pkg/ingester/tailer.go

```diff
@@ -0,0 +1,231 @@
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/labels"
+)
+
+const bufferSizeForTailResponse = 5
+
+type tailer struct {
+	id       uint32
+	orgID    string
+	matchers []*labels.Matcher
+	regexp   *regexp.Regexp
+
+	sendChan         chan *logproto.Stream
+	done             chan struct{}
+	ingesterQuitting chan struct{}
+
+	blockedAt      *time.Time
+	blockedMtx     sync.RWMutex
+	droppedStreams []*logproto.DroppedStream
+
+	conn logproto.Querier_TailServer
+}
+
+func newTailer(orgID, query, regex string, conn logproto.Querier_TailServer, ingesterQuitting chan struct{}) (*tailer, error) {
+	expr, err := logql.ParseExpr(query)
+	if err != nil {
+		return nil, err
+	}
+
+	matchers := expr.Matchers()
+	var re *regexp.Regexp
+
+	if regex != "" {
+		re, err = regexp.Compile(regex)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &tailer{
+		orgID:            orgID,
+		matchers:         matchers,
+		regexp:           re,
+		sendChan:         make(chan *logproto.Stream, bufferSizeForTailResponse),
+		conn:             conn,
+		droppedStreams:   []*logproto.DroppedStream{},
+		id:               generateUniqueID(orgID, query, regex),
+		done:             make(chan struct{}),
+		ingesterQuitting: ingesterQuitting,
+	}, nil
+}
+
+func (t *tailer) loop() {
+	var stream *logproto.Stream
+	var err error
+	var ok bool
+
+	ticker := time.NewTicker(3 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			err := t.conn.Context().Err()
+			if err != nil {
+				t.close()
+				return
+			}
+		case <-t.ingesterQuitting:
+			t.close()
+			return
+		case <-t.done:
+			return
+		case stream, ok = <-t.sendChan:
+			if !ok {
+				return
+			} else if stream == nil {
+				continue
+			}
+
+			// while sending new stream pop lined up dropped streams metadata for sending to querier
+			tailResponse := logproto.TailResponse{Stream: stream, DroppedStreams: t.popDroppedStreams()}
+			err = t.conn.Send(&tailResponse)
+			if err != nil {
+				level.Error(util.Logger).Log("Error writing to tail client", fmt.Sprintf("%v", err))
+				t.close()
+				return
+			}
+		}
+	}
+}
+
+func (t *tailer) send(stream logproto.Stream) {
+	if t.isClosed() {
+		return
+	}
+
+	// if we are already dropping streams due to blocked connection, drop new streams directly to save some effort
+	if blockedSince := t.blockedSince(); blockedSince != nil {
+		if blockedSince.Before(time.Now().Add(-time.Second * 15)) {
+			t.close()
+			close(t.sendChan)
+			return
+		}
+		t.dropStream(stream)
+		return
+	}
+
+	if t.regexp != nil {
+		// filter stream by regex from query
+		t.filterEntriesInStream(&stream)
+		if len(stream.Entries) == 0 {
+			return
+		}
+	}
+
+	select {
+	case t.sendChan <- &stream:
+	default:
+		t.dropStream(stream)
+	}
+}
+
+func (t *tailer) filterEntriesInStream(stream *logproto.Stream) {
+	filteredEntries := new([]logproto.Entry)
+	for _, entry := range stream.Entries {
+		if t.regexp.MatchString(entry.Line) {
```

Comment:
The new LogQL filter now has multiple regex. Do we want to support it ? this regex is meant to be deprecated.

Reply:
Please correct me if I am wrong, I see that its doing the same thing for query function here https://github.com/grafana/loki/blob/3b04bc7a2811f82ac738d8d11c05043412f2ebbd/pkg/ingester/instance.go#L112

Reply:
Look at this eval function :

https://github.com/grafana/loki/blob/3b04bc7a2811f82ac738d8d11c05043412f2ebbd/pkg/logql/ast.go#L63

Plus this is an AST tree so filter can contain another filter that can contain another filter .....